lib/iris/tests/unit/data_manager/test_DataManager.py

```diff
@@ -0,0 +1,602 @@
+
+    def test_lazy_with_real__realised_dtype_clearance(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        mask_array = ma.arange(size).reshape(shape)
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
+        dm = DataManager(lazy_array, realised_dtype=dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertEqual(dm._realised_dtype, dtype)
+        self.assertEqual(dm.dtype, dtype)
+        dm.replace(mask_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, dtype)
+        self.assertIs(dm._real_array, mask_array)
+        self.assertIs(dm.data, mask_array)
+
+    def test_real_with_lazy__realised_dtype_setter(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        mask_array = ma.masked_array(real_array) * 10
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
+        dm = DataManager(real_array)
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, dtype)
+        dm.replace(lazy_array, realised_dtype=dtype)
+        self.assertFalse(dm.has_real_data())
+        self.assertTrue(dm.has_lazy_data())
+        self.assertEqual(dm._realised_dtype, dtype)
+        self.assertEqual(dm.dtype, dtype)
+        self.assertIs(dm._lazy_array, lazy_array)
+        self.assertArrayEqual(dm.data, mask_array)
+
+    def test_real_with_real__realised_dtype_failure(self):
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        real_array = np.arange(size).reshape(shape)
+        dm = DataManager(real_array)
+        emsg = 'Cannot set realised dtype, no lazy data is available.'
+        with self.assertRaisesRegexp(ValueError, emsg):
+            dm.replace(real_array * 10, realised_dtype=real_array.dtype)
+        self.assertIs(dm._real_array, real_array)
+        self.assertArrayEqual(dm.data, real_array)
+
+
+class Test___copy__(tests.IrisTest):
+    def test(self):
+        dm = DataManager(np.array(0))
+        emsg = 'Shallow-copy of {!r} is not permitted.'
+        name = type(dm).__name__
+        with self.assertRaisesRegexp(copy.Error, emsg.format(name)):
+            copy.copy(dm)
+
+
+class Test___deepcopy__(tests.IrisTest):
+    def test(self):
+        dm = DataManager(np.array(0))
+        method = 'iris._data_manager.DataManager._deepcopy'
+        return_value = mock.sentinel.return_value
+        with mock.patch(method) as mocker:
+            mocker.return_value = return_value
+            result = copy.deepcopy(dm)
+            self.assertEqual(mocker.call_count, 1)
+        self.assertIs(result, return_value)
```

Comment:
Hmmm this is actually become some what painful ....

For this test `copy.deepcopy` populates the `memo` with the following example:
```python
   (Pdb) list
556  	        return_value = mock.sentinel.return_value
557  	        with mock.patch(method) as mocker:
558  	            mocker.return_value = return_value
559  	            result = copy.deepcopy(dm)
560  	            import pdb; pdb.set_trace()
561  ->	            self.assertEqual(mocker.call_count, 1)
562  	        self.assertIs(result, return_value)
563  	
564  	
565  	class Test_copy(tests.IrisTest):
566  	    def test(self):
(Pdb) p mocker.call_args
call({139777040548944: sentinel.return_value, 139777048606992: [DataManager(array(0))]})
```
If we want to introspect at this level, then I'm pretty much forced to implement `__eq__` and `__ne__` comparison methods on the `DataManager` ... that might be nice to have, but I'm not really motivated to do so just for a test. It may be that there is a genuine use case, say in `concatenate` or `merge` or somewhere else that might would be super handy, but I'm erring on the side of not doing it here and now .... thoughts @lbdreyer ?